README.md
# urlsafe-crypto
The [FreshFx](https://www.freshfx.at/) urlsafe-crypto lib.

## Components
* babel (with preset-env, plugin-add-jsdoc-properties, plugin-transform-object-reset-spread)
* eslint (with eslint-plugin-import)
* mocha, sinon & should
* conventional-changelog (use the angular cz style!)
* npm-run-all, onchange, release-it

## Setup

1. clone this repository
2. delete the `.git` folder
3. change those parts:
    * `index.js` -> adopt the module name
    * `package.json` -> edit the corresponding values
4. remove the .npmrc file if you don't want to use FreshFx private modules
5. change the settings in `.release-it.json` if you want to publish to npm
6. change the `template.hbs` file to fit your needs

**Attention:** to take advantage of the private npm modules, set the NPM_TOKEN as env variable

## Usage
see [package.json's](package.json#L6) "scripts" for all available commands

### commit
`npm run commit`

### start developing
`npm start`

### release
`npm run release`

## API Reference
encrypts, encodes and deflates a given object/string

Encryption Process:
 1. stringify data (optional)
 2. deflate
 3. encrypt
 4. deflate
 5. urlsafe base64 encode

Decryption Process:
 1. urlsafe base64 decode
 2. inflate
 3. decrypt
 4. inflate
 5. parse data (optional)

## Installation
`npm install --save @freshfx/urlsafe-crypto`

**Attention:** in order to install this package, either be a permitted user for this package, or create
a `.npmrc` in your projects root with following content (and set the NPM_TOKEN environment variable):
```txt
 //registry.npmjs.org/:_authToken=${NPM_TOKEN}
```

## Why should I use this?
and not f.e. (iron)[https://www.npmjs.com/package/iron]?
1. we don't need integrity (yet)
2. the generated string is shorter (30-50%) - good for URLs since the GET url length can be limited


* [@freshfx/urlsafe-crypto](#module_@freshfx/urlsafe-crypto)
    * [module.exports(encKey, isString)](#exp_module_@freshfx/urlsafe-crypto--module.exports) ⇒ <code>urlsafe-crypto</code> ⏏
        * _static_
            * [.encrypt](#module_@freshfx/urlsafe-crypto--module.exports.encrypt) ⇒ <code>Promise</code>
            * [.encryptSync](#module_@freshfx/urlsafe-crypto--module.exports.encryptSync) ⇒ <code>String</code>
            * [.decrypt](#module_@freshfx/urlsafe-crypto--module.exports.decrypt) ⇒ <code>String</code> \| <code>Object</code>
            * [.decryptSync](#module_@freshfx/urlsafe-crypto--module.exports.decryptSync) ⇒ <code>String</code> \| <code>Object</code>
        * _inner_
            * [~urlsafe-crypto](#module_@freshfx/urlsafe-crypto--module.exports..urlsafe-crypto) : <code>Object</code>

<a name="exp_module_@freshfx/urlsafe-crypto--module.exports"></a>

### module.exports(encKey, isString) ⇒ <code>urlsafe-crypto</code> ⏏
exports a init function which returns the specific methods
without the need of the encryption key

**Kind**: Exported function  
**Returns**: <code>urlsafe-crypto</code> - urlsafe-crypto functions  

| Param | Type | Description |
| --- | --- | --- |
| encKey | <code>String</code> | encryption key |
| isString | <code>Boolean</code> | set the default value for isString (f.e. if you only encrypt/decrypt strings) |

**Example**  
```js
import urlsafeCrypto from 'rbcp3-urlsafe-crypto'
const ENC_KEY = 'b6bad4846614652e7ead69df7337a7f4'
const crypto = urlsafeCrypto(ENC_KEY)

// prints the object
crypto.encrypt({an: 'object'}).then(crypto.decrypt).then(console.log)
// prints 'test', important: add the isString option to the decrypt function
crypto.encrypt('test').then(encryptedString => crypto.decrypt(encryptedString, true)).then(console.log)

const cryptoString = urlsafeCrypto(ENC_KEY, true)
// prints 'test'
cryptoString.encrypt('test').then(crypto.decrypt).then(console.log)
// prints the object, important: add the isString option to decrypt function
cryptoString.encrypt({an: 'object'}).then(encryptedObject => cryptoString.decrypt(encryptedObject, false)).then(console.log)
```
<a name="module_@freshfx/urlsafe-crypto--module.exports.encrypt"></a>

#### module.exports.encrypt ⇒ <code>Promise</code>
encrypts and encodes a given object or string

**Kind**: static constant of [<code>module.exports</code>](#exp_module_@freshfx/urlsafe-crypto--module.exports)  
**Returns**: <code>Promise</code> - Promise which resolves with the resulting String  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> \| <code>String</code> | the data to encrypt |
| encKey | <code>String</code> | Encryption Key, length must be 32 (256 Bit) |

**Example**  
```js
const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
// should print something like 'eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ'
encrypt({key: 'value'}, ENC_KEY).then(result => console.log(result))
```
<a name="module_@freshfx/urlsafe-crypto--module.exports.encryptSync"></a>

#### module.exports.encryptSync ⇒ <code>String</code>
encrypts and encodes a given object or string - synchronously

**Kind**: static constant of [<code>module.exports</code>](#exp_module_@freshfx/urlsafe-crypto--module.exports)  
**Returns**: <code>String</code> - the resulting string  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> \| <code>String</code> | the data to encrypt |
| encKey | <code>String</code> | Encryption Key, length must be 32 (256 Bit) |

**Example**  
```js
const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
// should print something like 'eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ'
console.log(encrypt({key: 'value'}, ENC_KEY))
```
<a name="module_@freshfx/urlsafe-crypto--module.exports.decrypt"></a>

#### module.exports.decrypt ⇒ <code>String</code> \| <code>Object</code>
decodes and decrypts a given string

**Kind**: static constant of [<code>module.exports</code>](#exp_module_@freshfx/urlsafe-crypto--module.exports)  
**Returns**: <code>String</code> \| <code>Object</code> - the resulting string/object  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| string | <code>String</code> |  | the data to decode/decrypt |
| encKey | <code>String</code> |  | Encryption Key, length must be 32 (256 Bit) |
| toString | <code>Boolean</code> | <code>false</code> | indicates if the result should be converted into a string or object |

**Example**  
```js
const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
// prints the object'{"key": "value"}'
decrypt('eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ', ENC_KEY)
  .then(console.log)

// prints 'tests'
encrypt('test', ENC_KEY).then(result => decrypt(result, ENC_KEY, true)).then(console.log)
```
<a name="module_@freshfx/urlsafe-crypto--module.exports.decryptSync"></a>

#### module.exports.decryptSync ⇒ <code>String</code> \| <code>Object</code>
decodes and decrypts a given string - synchronously

**Kind**: static constant of [<code>module.exports</code>](#exp_module_@freshfx/urlsafe-crypto--module.exports)  
**Returns**: <code>String</code> \| <code>Object</code> - the resulting string/object  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| string | <code>String</code> |  | the data to decode/decrypt |
| encKey | <code>String</code> |  | Encryption Key, length must be 32 (256 Bit) |
| toString | <code>Boolean</code> | <code>false</code> | indicates if the result should be converted into a string or object |

**Example**  
```js
const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
// prints the object'{"key": "value"}'
console.log(decrypt('eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ', ENC_KEY))

// prints 'tests'
console.log(decrypt(encrypt('test', ENC_KEY), ENC_KEY, true))
```
<a name="module_@freshfx/urlsafe-crypto--module.exports..urlsafe-crypto"></a>

#### module.exports~urlsafe-crypto : <code>Object</code>
**Kind**: inner typedef of [<code>module.exports</code>](#exp_module_@freshfx/urlsafe-crypto--module.exports)  
**Properties**

| Name | Type |
| --- | --- |
| encrypt | <code>function</code> | 
| encryptSync | <code>function</code> | 
| decrypt | <code>function</code> | 
| decryptSync | <code>function</code> | 


* * *

&copy; 2021 FreshFx <npm@freshfx.at>.


CHANGELOG.md


# [6.0.0](https://github.com/freshfx/rbcp3-urlsafe-crypto/compare/v5.0.0...v6.0.0) (2022-10-13)

## 6.0.0 (2022-10-13)

* chore: update deps ([d14f130](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/d14f130))



## 5.0.0 (2021-10-18)

* chore: release v5.0.0 ([744302a](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/744302a))
* feat: upgaded to node v16/npm v8 ([6b983f5](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/6b983f5))


### BREAKING CHANGE

* node.js/npm upgraded


## 4.2.0 (2021-09-09)

* chore: release v4.2.0 ([fbf8976](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/fbf8976))
* feat: update .gitignore ([60ec473](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/60ec473))
* feat: update .gitignore ([c2f2155](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/c2f2155))



## <small>4.1.4 (2021-09-09)</small>

* chore: release v4.1.4 ([dacdd2e](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/dacdd2e))
* chore(package.json): updated deps ([b323c83](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/b323c83))



## <small>4.1.3 (2021-07-27)</small>

* chore: release v4.1.3 ([7db2ee8](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/7db2ee8))
* chore: upgraded dev platform ([93c562e](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/93c562e))



## 4.1.0 (2020-02-07)

* chore: release v4.1.0 ([66ed48b](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/66ed48b))
* chore: update deps ([e2effde](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/e2effde))
* chore(deps): bump handlebars from 4.1.2 to 4.5.3 ([f9881c0](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/f9881c0))



## 4.0.0 (2019-06-05)

* chore: release v4.0.0 ([8d672b0](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/8d672b0))
* chore(package.json): updated deps ([2d7b655](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/2d7b655))



## 3.1.0 (2019-01-17)

* chore: release v3.1.0 ([eff67d1](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/eff67d1))
* chore(package.json): updated deps ([a157960](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/a157960))



## 3.0.0 (2019-01-03)

* chore: release v3.0.0 ([80f34d8](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/80f34d8))
* chore(nvm): updated to v11 ([f03bfbf](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/f03bfbf))
* chore(package.json): updated deps ([68331d8](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/68331d8))
* feat(crypt/zlib): changed compression to BEST_SPEED (better performance) ([281861c](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/281861c))



## <small>2.0.3 (2018-11-27)</small>

* chore: release v2.0.3 ([f7efe4a](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/f7efe4a))
* chore(package.json): updated dev deps ([cb29676](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/cb29676))



## <small>2.0.2 (2018-10-11)</small>

* chore: release v2.0.2 ([4c9d4ef](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/4c9d4ef))
* docs(installation): added npm info ([5d5d1b0](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/5d5d1b0))



## <small>2.0.1 (2018-10-11)</small>

* chore: release v2.0.1 ([748d345](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/748d345))
* fix(package.json): removed private flag for npm publish ([a32d188](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/a32d188))



## 2.0.0 (2018-10-11)

* chore: release v2.0.0 ([349486e](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/349486e))
* feat: migrated to module-boilerplate ([651420b](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/651420b))



## 1.1.0 (2018-10-02)

* refactor: updated deps, changed to async await ([60e4d5e](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/60e4d5e))
* feat: updated node to v10, fixed test ([68ef679](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/68ef679))
* chore: upate jsdoc deps ([7e605ca](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/7e605ca))
* docs: changed install url from ssh to https ([2546091](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/2546091))



## <small>1.0.1 (2017-10-17)</small>

* chore(package.json): updated dependencies ([b58b19e](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/b58b19e))
* docs: updated import statement ([e925afa](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/e925afa))



## 1.0.0 (2017-09-13)

* docs: added installation instructions ([076e656](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/076e656))
* fix: fixed main script, added preinstall hook to run build ([694fc42](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/694fc42))
* fix: removed package.json from dist directory, removed copy script ([0c6e27c](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/0c6e27c))
* chore: added dist dir ([c94158b](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/c94158b))
* refactor: updated example, added missing tests ([e95b6ff](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/e95b6ff))
* feat: initialized repo, added code/tests etc. ([74ba520](https://github.com/freshfx/rbcp3-urlsafe-crypto/commit/74ba520))

test/index.js
import {randomBytes} from 'crypto'

import deepFreeze from 'deep-freeze'
import should from 'should'

import * as urlsafeCrypto from '../src'

import asPromiseTests from './lib/as-promise'
import base64UrlsafeTests from './lib/base64-urlsafe'
import jsonTests from './lib/json'
import zlibTests from './lib/zlib'
import cryptTests from './lib/crypt'

asPromiseTests()
base64UrlsafeTests()
jsonTests()
zlibTests()
cryptTests()


describe('urlsafe-crypto', () => {
  const ENC_KEY = randomBytes(32)
  const inputObject = deepFreeze({an: 'object'})
  const inputString = 'this is a string'
  let encryptedObject = null
  let encryptedString = null

  describe('encrypt', () => {
    it('should be exported', () => {
      should.exist(urlsafeCrypto.encrypt)
      urlsafeCrypto.encrypt.should.be.a.Function()
    })

    it('should encrypt an object', () =>
      urlsafeCrypto.encrypt(inputObject, ENC_KEY).then(result => {
        encryptedObject = result
        result.should.be.a.String()
        result.length.should.be.above(20)
        const regex = /^[a-zA-Z0-9_-]+$/
        regex.test(result).should.be.True()
      })
    )

    it('should encrypt a string', () =>
      urlsafeCrypto.encrypt(inputString, ENC_KEY).then(result => {
        encryptedString = result
        result.should.be.a.String()
        result.length.should.be.above(20)
        const regex = /^[a-zA-Z0-9_-]+$/
        regex.test(result).should.be.True()
      })
    )
  })

  describe('encryptSync', () => {
    it('should be exported', () => {
      should.exist(urlsafeCrypto.encryptSync) // eslint-disable-line no-sync
      urlsafeCrypto.encryptSync.should.be.a.Function() // eslint-disable-line no-sync
    })

    it('should encrypt an object', () => {
      const result = urlsafeCrypto.encryptSync(inputObject, ENC_KEY) // eslint-disable-line no-sync
      result.should.be.a.String()
      result.length.should.be.above(20)
      const regex = /^[a-zA-Z0-9_-]+$/
      regex.test(result).should.be.True()
    })

    it('should encrypt a string', () => {
      const result = urlsafeCrypto.encryptSync(inputString, ENC_KEY) // eslint-disable-line no-sync
      result.should.be.a.String()
      result.length.should.be.above(20)
      const regex = /^[a-zA-Z0-9_-]+$/
      regex.test(result).should.be.True()
    })
  })


  describe('decrypt', () => {
    it('should be exported', () => {
      should.exist(urlsafeCrypto.encrypt)
      urlsafeCrypto.encrypt.should.be.a.Function()
    })

    it('should decrypt an object', () =>
      urlsafeCrypto.decrypt(encryptedObject, ENC_KEY).then(result => {
        result.should.eql(inputObject)
      })
    )

    it('should decrypt a string', () =>
      urlsafeCrypto.decrypt(encryptedString, ENC_KEY, true).then(result => {
        result.should.equal(inputString)
      })
    )
  })


  describe('decryptSync', () => {
    it('should be exported', () => {
      should.exist(urlsafeCrypto.decryptSync) // eslint-disable-line no-sync
      urlsafeCrypto.decryptSync.should.be.a.Function() // eslint-disable-line no-sync
    })

    it('should decrypt an object', () => {
      urlsafeCrypto.decryptSync(encryptedObject, ENC_KEY).should.eql(inputObject) // eslint-disable-line no-sync
    })

    it('should decrypt a string', () => {
      urlsafeCrypto.decryptSync(encryptedString, ENC_KEY, true).should.equal(inputString) // eslint-disable-line no-sync
    })
  })


  describe('export default', () => {
    it('should be a Function', () => {
      urlsafeCrypto.default.should.be.a.Function()
    })

    it('should throw if encKey is missing or invalid', () => {
      const doIt = key => () => urlsafeCrypto.default(key)
      doIt().should.throw('encryption key is missing')
      doIt({}).should.throw('a Buffer or String is required')
      doIt('asdf').should.throw('invalid encryption key length (is: 4 bytes, should be: 32 bytes)')
      doIt(randomBytes(32)).should.not.throw()
    })

    it('should return the crypto functions', () => { // eslint-disable-line max-statements
      const funcs = urlsafeCrypto.default(randomBytes(32))
      should.exist(funcs.decrypt)
      funcs.decrypt.should.be.a.Function()
      should.exist(funcs.decryptSync) // eslint-disable-line no-sync
      funcs.decryptSync.should.be.a.Function() // eslint-disable-line no-sync
      should.exist(funcs.encrypt)
      funcs.encrypt.should.be.a.Function()
      should.exist(funcs.encryptSync) // eslint-disable-line no-sync
      funcs.encryptSync.should.be.a.Function() // eslint-disable-line no-sync

      const doIt = () => () => funcs.encrypt(Buffer.from([]))
      doIt().should.not.throw()
    })

    it('should use isString if set', done => {
      const funcs = urlsafeCrypto.default(randomBytes(32), true)
      const input = 'whaaaaat'
      funcs
        .encrypt(input)
        .then(result => {
          result.should.not.equal(input)
          return Promise.resolve(result)
        })
        .then(funcs.decrypt)
        .then(result => {
          result.should.equal(input)
          const encryptedSync = funcs.encryptSync(input) // eslint-disable-line no-sync
          encryptedSync.should.not.equal(input)
          funcs.decryptSync(encryptedSync).should.equal(input) // eslint-disable-line no-sync
          done()
        })
        .catch(done)
    })

    it('should encrypt/decrypt objects', done => {
      const funcs = urlsafeCrypto.default(randomBytes(32))
      const input = {
        a: {
          nested: 'one'
        },
        an: 'object',
        key: 'value',
        properties: 'or',
        with: 'some'
      }
      funcs
        .encrypt(input)
        .then(result => {
          result.should.not.eql(input)
          return Promise.resolve(result)
        })
        .then(funcs.decrypt)
        .then(result => {
          result.should.eql(input)
          const encryptedSync = funcs.encryptSync(input) // eslint-disable-line no-sync
          encryptedSync.should.not.equal(input)
          funcs.decryptSync(encryptedSync).should.eql(input) // eslint-disable-line no-sync
          done()
        })
        .catch(done)
    })

    it('should encrypt/decrypt both', done => {
      const ENC_KEY = randomBytes(32)
      const funcs = urlsafeCrypto.default(ENC_KEY)
      const funcsString = urlsafeCrypto.default(ENC_KEY, true)
      const input = {key: 'value'}
      const inputString = 'testString'

      Promise.all([
        funcs.encrypt(input),
        funcs.encrypt(inputString),
        funcsString.encrypt(input),
        funcsString.encrypt(inputString)
      ])
      .then(([encryptedObject, encryptedString, encryptedObject2, encryptedString2]) =>
        Promise.all([
          funcs.decrypt(encryptedObject2),
          funcs.decrypt(encryptedString2, true),
          funcsString.decrypt(encryptedObject, false),
          funcsString.decrypt(encryptedString)
        ])
      )
      .then(([decryptedObject, decryptedString, decryptedObject2, decryptedString2]) => {
        decryptedObject.should.eql(decryptedObject2)
        decryptedObject.should.eql(input)
        decryptedString.should.equal(decryptedString2)
        decryptedString.should.equal(inputString)

        funcsString.decryptSync(funcs.encryptSync(input), false).should.eql(input) // eslint-disable-line no-sync
        funcsString.decryptSync(funcs.encryptSync(inputString)).should.equal(inputString) // eslint-disable-line no-sync
        funcs.decryptSync(funcsString.encryptSync(input)).should.eql(input) // eslint-disable-line no-sync
        funcs.decryptSync(funcsString.encryptSync(inputString), true).should.equal(inputString) // eslint-disable-line no-sync
        done()
      })
      .catch(done)
    })
  })
})


test/lib/json.js
import should from 'should'
import sinon from 'sinon'

import * as json from '../../src/lib/json'

require('should-sinon')

export default () =>
  describe('json', () => {
    describe('stringify', () => {
      it('should be exported', () => {
        should.exist(json.stringify)
        json.stringify.should.be.a.Function()
      })

      it('should call JSON.stringify', () => {
        const spy = sinon.spy(JSON, 'stringify')
        const input = {key: 'test'}
        json.stringify(input)
        spy.callCount.should.equal(1)
        spy.firstCall.args[0].should.equal(input)
        spy.restore()
      })
    })

    describe('stringifyP', () => {
      it('should be exported', () => {
        should.exist(json.stringifyP)
        json.stringifyP.should.be.a.Function()
      })

      it('should stringify the data', done => {
        const input = {key: 'test'}
        const expected = '{"key":"test"}'
        json.stringifyP(input)
          .then(result => {
            result.should.equal(expected)
            done()
          })
          .catch(done)
      })
    })


    describe('parse', () => {
      it('should be exported', () => {
        should.exist(json.parse)
        json.parse.should.be.a.Function()
      })

      it('should call JSON.parse', () => {
        const spy = sinon.spy(JSON, 'parse')
        const input = '{"key": "test"}'
        json.parse(input)
        spy.callCount.should.equal(1)
        spy.firstCall.args[0].should.equal(input)
        spy.restore()
      })
    })

    describe('parseP', () => {
      it('should be exported', () => {
        should.exist(json.parseP)
        json.parseP.should.be.a.Function()
      })

      it('should stringify the data', done => {
        const expected = {key: 'test'}
        const input = '{"key":"test"}'
        json.parseP(input)
          .then(result => {
            result.should.eql(expected)
            done()
          })
          .catch(done)
      })
    })

    describe('export default', () => {
      it('should export stringify', () => {
        should.exist(json.default.stringify)
        json.stringify.should.equal(json.default.stringify)
      })

      it('should export stringifyP', () => {
        should.exist(json.default.stringifyP)
        json.stringifyP.should.equal(json.default.stringifyP)
      })

      it('should export parse', () => {
        should.exist(json.default.parse)
        json.parse.should.equal(json.default.parse)
      })

      it('should export parseP', () => {
        should.exist(json.default.parseP)
        json.parseP.should.equal(json.default.parseP)
      })
    })
  })


src/index.js
import {Z_BEST_SPEED} from 'zlib'

import * as crypt from './lib/crypt'
import {parse, parseP, stringify, stringifyP} from './lib/json'
import {deflate, deflateSync, inflate, inflateSync} from './lib/zlib'
import {encode, encodeP, decode, decodeP} from './lib/base64-urlsafe'

/**
 * encrypts, encodes and deflates a given object/string
 *
 * Encryption Process:
 *  1. stringify data (optional)
 *  2. deflate
 *  3. encrypt
 *  4. deflate
 *  5. urlsafe base64 encode
 *
 * Decryption Process:
 *  1. urlsafe base64 decode
 *  2. inflate
 *  3. decrypt
 *  4. inflate
 *  5. parse data (optional)
 *
 * ## Installation
 * `npm install --save @freshfx/urlsafe-crypto`
 *
 * **Attention:** in order to install this package, either be a permitted user for this package, or create
 * a `.npmrc` in your projects root with following content (and set the NPM_TOKEN environment variable):
 * ```txt
 *  //registry.npmjs.org/:_authToken=${NPM_TOKEN}
 * ```
 *
 * ## Why should I use this?
 * and not f.e. (iron)[https://www.npmjs.com/package/iron]?
 * 1. we don't need integrity (yet)
 * 2. the generated string is shorter (30-50%) - good for URLs since the GET url length can be limited
 *
 * @module @freshfx/urlsafe-crypto
 */

/**
 * encrypts and encodes a given object or string
 * @example
 * const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
 * // should print something like 'eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ'
 * encrypt({key: 'value'}, ENC_KEY).then(result => console.log(result))
 * @param  {Object|String} data the data to encrypt
 * @param  {String} encKey Encryption Key, length must be 32 (256 Bit)
 * @return {Promise} Promise which resolves with the resulting String
 */
export const encrypt = async (data, encKey) => {
  if (typeof data === 'object') {
    data = await stringifyP(data)
  }

  const deflatedData = await deflate(data, {level: Z_BEST_SPEED})
  const encryptedDeflatedData = await crypt.encryptP(deflatedData, encKey)
  const deflatedEncryptedDeflatedData = await deflate(encryptedDeflatedData, {level: Z_BEST_SPEED})
  const encodedDeflatedEncryptedDeflatedData = await encodeP(deflatedEncryptedDeflatedData)
  return encodedDeflatedEncryptedDeflatedData
}

/**
 * encrypts and encodes a given object or string - synchronously
 * @example
 * const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
 * // should print something like 'eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ'
 * console.log(encrypt({key: 'value'}, ENC_KEY))
 * @param  {Object|String} data the data to encrypt
 * @param  {String} encKey Encryption Key, length must be 32 (256 Bit)
 * @return {String} the resulting string
 */
export const encryptSync = (data, encKey) => {
  if (typeof data === 'object') {
    data = stringify(data)
  }
  const deflatedData = deflateSync(data)
  const encryptedDeflatedData = crypt.encrypt(deflatedData, encKey)
  const deflatedEncryptedDeflatedData = deflateSync(encryptedDeflatedData, {level: Z_BEST_SPEED})
  const encodedDeflatedEncryptedDeflatedData = encode(deflatedEncryptedDeflatedData)
  return encodedDeflatedEncryptedDeflatedData
}

/**
 * decodes and decrypts a given string
 * @example
 * const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
 * // prints the object'{"key": "value"}'
 * decrypt('eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ', ENC_KEY)
 *   .then(console.log)
 *
 * // prints 'tests'
 * encrypt('test', ENC_KEY).then(result => decrypt(result, ENC_KEY, true)).then(console.log)
 * @param {String} string the data to decode/decrypt
 * @param {String} encKey Encryption Key, length must be 32 (256 Bit)
 * @param {Boolean} toString=false indicates if the result should be converted into a string or object
 * @return {String|Object} the resulting string/object
 */
export const decrypt = async (string, encKey, toString = false) => {
  const deflatedEncryptedDeflatedData = await decodeP(string)
  const encryptedDeflatedData = await inflate(deflatedEncryptedDeflatedData)
  const deflatedData = await crypt.decryptP(encryptedDeflatedData, encKey)
  const data = (await inflate(deflatedData)).toString()

  if (toString === true) {
    return data
  }
  return parseP(data)
}

/**
 * decodes and decrypts a given string - synchronously
 * @example
 * const ENC_KEY = '6b7beea8ef24f7ee89e153387db8f04f'
 * // prints the object'{"key": "value"}'
 * console.log(decrypt('eJwNzAcBA0EIADBLbI66YfqX8I2ACCUcg2dSRseFKzIYpAtXM_yWN9_sDVwk9lxORTkS6op3FHdJdOmqnTI1rbrJP-FnD-wDMqcaoQ', ENC_KEY))
 *
 * // prints 'tests'
 * console.log(decrypt(encrypt('test', ENC_KEY), ENC_KEY, true))
 * @param {String} string the data to decode/decrypt
 * @param {String} encKey Encryption Key, length must be 32 (256 Bit)
 * @param {Boolean} toString=false indicates if the result should be converted into a string or object
 * @return {String|Object} the resulting string/object
 */
export const decryptSync = (string, encKey, toString = false) => {
  const deflatedEncryptedDeflatedData = decode(string)
  const encryptedDeflatedData = inflateSync(deflatedEncryptedDeflatedData)
  const deflatedData = crypt.decrypt(encryptedDeflatedData, encKey)
  const data = inflateSync(deflatedData).toString()

  if (toString === true) {
    return data
  }
  return parse(data)
}


/**
 * @typedef {Object} urlsafe-crypto
 * @property {Function} encrypt
 * @property {Function} encryptSync
 * @property {Function} decrypt
 * @property {Function} decryptSync
 */


/**
 * exports a init function which returns the specific methods
 * without the need of the encryption key
 * @example
 * import urlsafeCrypto from 'rbcp3-urlsafe-crypto'
 * const ENC_KEY = 'b6bad4846614652e7ead69df7337a7f4'
 * const crypto = urlsafeCrypto(ENC_KEY)
 *
 * // prints the object
 * crypto.encrypt({an: 'object'}).then(crypto.decrypt).then(console.log)
 * // prints 'test', important: add the isString option to the decrypt function
 * crypto.encrypt('test').then(encryptedString => crypto.decrypt(encryptedString, true)).then(console.log)
 *
 * const cryptoString = urlsafeCrypto(ENC_KEY, true)
 * // prints 'test'
 * cryptoString.encrypt('test').then(crypto.decrypt).then(console.log)
 * // prints the object, important: add the isString option to decrypt function
 * cryptoString.encrypt({an: 'object'}).then(encryptedObject => cryptoString.decrypt(encryptedObject, false)).then(console.log)
 * @param {String} encKey - encryption key
 * @param {Boolean} isString - set the default value for isString (f.e. if you only encrypt/decrypt strings)
 * @return {urlsafe-crypto} urlsafe-crypto functions
 */
export default (encKey, isString = false) => {
  const defaultIsString = isString
  if (!encKey) {
    throw new Error('encryption key is missing')
  }
  if (typeof encKey === 'string') {
    encKey = Buffer.from(encKey)
  }
  if (!Buffer.isBuffer(encKey)) {
    throw new Error('a Buffer or String is required')
  }
  if (encKey.length !== 32) {
    throw new Error(`invalid encryption key length (is: ${encKey.length} bytes, should be: 32 bytes)`)
  }
  return {
    decrypt: (data, isString = defaultIsString) => decrypt(data, encKey, isString),
    decryptSync: (data, isString = defaultIsString) => decryptSync(data, encKey, isString),
    encrypt: data => encrypt(data, encKey),
    encryptSync: data => encryptSync(data, encKey)
  }
}
